pkg/controller/core/clusterqueue_controller_test.go

```diff
@@ -489,3 +487,82 @@ func TestRecordResourceMetrics(t *testing.T) {
 		})
 	}
 }
+
+func TestStartSnapshot(t *testing.T) {
+	cqName := "test-cq"
+	lqName := "test-lq"
+	const lowPrio, highPrio = 0, 100
+	defaultWls := &kueue.WorkloadList{
+		Items: []kueue.Workload{
+			*utiltesting.MakeWorkload("one", "").Queue(lqName).Priority(highPrio).Obj(),
+			*utiltesting.MakeWorkload("two", "").Queue(lqName).Priority(lowPrio).Obj(),
+		},
+	}
+	testCases := map[string]struct {
+		queueVisibilityUpdateInterval        time.Duration
+		queueVisibilityClusterQueuesMaxCount int32
+		wantPendingWorkloadsStatus           *kueue.ClusterQueuePendingWorkloadsStatus
+	}{
+		"taking snapshot of cluster queue is disabled": {},
+		"taking snapshot of cluster queue is enabled": {
+			queueVisibilityClusterQueuesMaxCount: 2,
+			queueVisibilityUpdateInterval:        10 * time.Millisecond,
+			wantPendingWorkloadsStatus: &kueue.ClusterQueuePendingWorkloadsStatus{
+				Head: []kueue.ClusterQueuePendingWorkload{
+					{Name: "one"}, {Name: "two"},
+				},
+			},
+		},
+		"verify the head of pending workloads when the number of pending workloads exceeds MaxCount": {
+			queueVisibilityClusterQueuesMaxCount: 1,
+			queueVisibilityUpdateInterval:        10 * time.Millisecond,
+			wantPendingWorkloadsStatus: &kueue.ClusterQueuePendingWorkloadsStatus{
+				Head: []kueue.ClusterQueuePendingWorkload{
+					{Name: "one"},
+				},
+			},
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			cq := utiltesting.MakeClusterQueue(cqName).
+				QueueingStrategy(kueue.StrictFIFO).Obj()
+			lq := utiltesting.MakeLocalQueue(lqName, "").
+				ClusterQueue(cqName).Obj()
+			ctx := context.Background()
+
+			cl := utiltesting.NewClientBuilder().WithLists(defaultWls).WithObjects(lq, cq).WithStatusSubresource(lq, cq).
+				Build()
+			cCache := cache.New(cl)
+			qManager := queue.NewManager(cl, cCache)
+			if err := qManager.AddClusterQueue(ctx, cq); err != nil {
+				t.Fatalf("Inserting clusterQueue in manager: %v", err)
+			}
+			if err := qManager.AddLocalQueue(ctx, lq); err != nil {
+				t.Fatalf("Inserting localQueue in manager: %v", err)
+			}
+
+			r := NewClusterQueueReconciler(
+				cl,
+				qManager,
+				cCache,
+				WithQueueVisibilityUpdateInterval(tc.queueVisibilityUpdateInterval),
+				WithQueueVisibilityClusterQueuesMaxCount(tc.queueVisibilityClusterQueuesMaxCount),
+			)
+
+			if err := startSnapshotWithTimeout(ctx, r); err != nil {
+				t.Errorf("error startSnapshotWithTimeout: %v", err)
+			}
+
+			if diff := cmp.Diff(tc.wantPendingWorkloadsStatus, r.getWorkloadsStatus(cq), cmpopts.IgnoreFields(kueue.ClusterQueuePendingWorkloadsStatus{}, "LastChangeTime")); len(diff) != 0 {
```

Comment:
I think before comparing we need to wait until the snapshot is computed. Please verify (without merging) the test passes when the updateInterval=2s.

Reply:
I checked tests with that interval, the test passes

Reply:
Probably, you should be able to repro this by injecting a sleep before `snapshotsQueue.Get`